mitiq/pt/pt.py

```diff
@@ -0,0 +1,61 @@
+# Copyright (C) Unitary Fund
+#
+# This source code is licensed under the GPL license (v3) found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Callable, List, Optional, Tuple, Union
+
+import cirq
+import numpy as np
+
+from mitiq import QPROGRAM, Executor, Observable, QuantumResult
+from mitiq.interface import atomic_converter
+
+
+def execute_with_pauli_twirling(
+    circuit: QPROGRAM,
+    executor: Union[Executor, Callable[[QPROGRAM], QuantumResult]],
+    observable: Optional[Observable] = None,
+    *,
+    num_circuits: int = 100,
+) -> float:
+    executor = (
+        executor if isinstance(executor, Executor) else Executor(executor)
+    )
+    twirled_circuits = pauli_twirled_circuits(circuit, num_circuits)
+    expvals = executor.evaluate(twirled_circuits, observable)
+    return np.average(expvals)
+
+
+def pauli_twirled_circuits(
+    circuit: QPROGRAM, num_circuits: int
+) -> List[QPROGRAM]:
+    return [add_paulis(circuit) for _ in range(num_circuits)]
+
+
+Paulis = Union[cirq.IdentityGate, cirq.XPowGate, cirq.YPowGate, cirq.ZPowGate]
+
+
+def sample_paulis() -> Tuple[Paulis, Paulis, Paulis, Paulis]:
+    # P, Q, R, S
+    return cirq.ops.X, cirq.ops.I, cirq.ops.X, cirq.ops.X
+
+
+@atomic_converter
+def add_paulis(circuit: cirq.Circuit) -> cirq.Circuit:
+    return circuit.map_operations(twirl_CNOT)
+
+
+def twirl_CNOT(op: cirq.Operation) -> cirq.OP_TREE:
+    if op.gate != cirq.CNOT:
```

Comment:
We could twirl CZ gates too, in a similar way.
If so `twirl_CNOT`--> `twirl_CNOT_and_CZ` or it could be a new function.

Reply:
Yeah, Cirq's `map_operations` method makes this a pretty straightforward change. I think having an additional function for $\mathrm{C}Z$ (that could be combined with this one) would be great.

Reply:
@Aaron-Robertson I noticed you combined the `twirl_CNOT` function and the proposed `twirl_CZ` function. Seeing as this is a new technique that we don't fully understand, I think it might be best to keep them separate for now. Do you have a preference to keeping them as a single function?

Reply:
From a code perspective I like the single function, since we get code reusability into the future. As a new `mitiq` technique I certainly see where you're coming from. How about we keep what we have and hide the general case, while exposing the specific?

Reply:
> How about we keep what we have and hide the general case, while exposing the specific?

What do you mean by this? Can you clarify?

Throwing another idea out there, we could have a function `generate_twirl_func(gate) -> Callable[[cirq.Operation], cirq.OP_TREE]` which you could then use to create `twirl_CNOT = generate_twirl_func(cirq.CNOT)`, and `twirl_CZ = generate_twirl_func(cirq.CZ)`. Don't love this approach, but don't hate it either.

Reply:
That's essentially what I meant, but I like the simpler call into the existing `_twirl_two_qubit_gate`:
```
def _twirl_two_qubit_gate(op: cirq.Operation) -> cirq.OP_TREE:...

def twirl_CZ() -> cirq.OP_TREE:
    return _twirl_two_qubit_gate(cirq.CZ)
    
def twirl_CNOT() -> cirq.OP_TREE:
    return _twirl_two_qubit_gate(cirq.CNOT)
```